### PR TITLE
Some changes to better support newer versions for RHEL/CentOS

### PR DESCRIPTION
**Pull request by rbrunckhorst**

### CODE REVIEW

File: packaging/cellcc-el7.spec
Comment (adeason) [inline packaging/cellcc-el7.spec]: hmm, when we spoke earlier and you mentioned putting 'git log' data into the specfile changelog, I thought we were talking about generating the %changelog data as part of the build process. I definitely do not want to keep 'git log' data checked in to a file like this.

but if you wanted to have something in './packaging/rules srpm7' generate this at build time, that would be okay. that is, we keep a cellcc-el7.spec.in file, and run some script like './generate-changelog cellcc-el7.spec.in > cellcc-el7.spec' (the .spec.in file is in git, but the .spec file is not). or if it's possible for the %prep or %build steps in the specfile to do that, that's another way (but I assume it's not possible to alter a spec file during those steps?)

File: INSTALL
Comment (adeason) [inline INSTALL]: these aren't really requirements, but one way to satisfy the other requirements (some perl modules and remctl, I think?). I think this belongs in the "RPM" section

File: packaging/rules
Comment (adeason) [inline packaging/rules:9]: I prefer to keep separate spec files around to avoid complicated %if checks inside the spec, but if it really doesn't need any changes for rhel8 and modern fedora, then that's perfectly fine.

even with separate specfiles, I suppose if we wanted, we could just detect what platform we're on and pick a spec file from that.

File: packaging/rules
Comment (adeason) [inline packaging/rules]: I do want to mention in here what platforms the packaging "supports" (so someone isn't surprised if this doesn't work on rhel6 or suse or whatever). but this may be better as a separate note after all of these, like "rpm packaging currently supported for RHEL/CentOS 7-8, and Fedora 33-34"

**Review by adeason — CHANGES_REQUESTED** [review]
I'm going to try pushing some suggested changes for the minor stuff in a moment, but I won't touch the changelog stuff

**Comment by adeason** [comment]
Take a look at the second commit; do those changes look alright to you? If you're okay with that, I can merge just the second commit right away and we can have a separate PR for the changelog stuff, if you like.

File: INSTALL
Comment (adeason) [inline INSTALL]: I nearly forgot; do we need epel on rhel8 ? your original text mentioned just the powertools repo for rhel8 specifically; I assumed we'd still need also need epel enabled separately, but is that not true?

**Comment by rbrunckhorst** [comment]
Yes, correct.
EPEL is still needed for rhel8.
Only on modern fedora there is no need to specify extra needed repos.

**Comment by rbrunckhorst** [comment]
- Regarding changelog:
   I thought it would be useful to have the changelog info also in the 
package. I was use to it when working with the openSuSE openAFS src 
package where those changes are included in the spec-file.
   That was also true in the openafs git repo for Redhat in the past:
   * Wed Dec 12 2007  Simon Wilkinson ***@***.***> 1.4.5
   * ...
   * Sat Sep 10 2005  Derek Atkins ***@***.***> 1.4
   But is was not maintained anymore in newer versions.
   The tool that I have used (generate-changelog) needs to be installed 
to be able to generate the changelog in a spec-file format. So we would 
add another dependency.
   It would be possible to use pure 'git log' commands with some format 
and maybe string manipulation to get a similar output, but I haven't 
looked into that.
   If your prefer not to have this info in the spec-file we can skip it, 
I think the effort is not worth it.

- Regarding separate spec-files
   I have no problems with that. I only thought it is more simple and 
currently not needed with the mentioned platforms.
   So if you like I can take care to create the  needed spec-files for 
those platforms.

- Missing requirements
   Today I have tried again to build and install cellcc on fresh 
installed platforms: rhel7, rhel8, fedora34
   I figure out that some packages are not installed automatically with 
the current spec-file. I was able to fix that by adding the following 
lines in the spec-file.
   (This was working for all those platforms.)
   - Requires: perl-DBIx-Simple
   - Requires: perl-DateTime-Format-MySQL
   - Requires: perl-Log-Log4perl
   - Requires: perl-Parallel-ForkManager
   - Requires: perl-Test-Spelling
   - Requires: perl-Test-Pod
   - Requires: perl-Filesys-Df
   I can create a new PR if you like after the current one is merged.

- Other things that come up
   There are some preparations needed (not mentioned in the INSTALL 
file) to be able to build cellcc
   - yum install "perl(ExtUtils::MakeMaker)" rpm-build git
     Question: Should this be mentioned in the INSTALL file?
   - install afs-command
     Question: Should we add the repo address 
https://github.com/openafs-contrib/afs-command in the INSTALL file?
     Or even better, should we add some packaging to afs-command similar 
like in cellcc? With that this could be addressed with an additional 
Requires section in the spec-file for cellcc.

BTW: I have also tried to build/install cellcc on rhel6: This was 
successfully done with the same procedure like for the other platforms.

/Ralf

**Comment by adeason** [comment]
On Tue, 17 Aug 2021 01:20:48 -0700
Ralf Brunckhorst ***@***.***> wrote:

> - Regarding changelog:
>    I thought it would be useful to have the changelog info also in the 
> package. I was use to it when working with the openSuSE openAFS src 
> package where those changes are included in the spec-file.
>    That was also true in the openafs git repo for Redhat in the past:
>    * Wed Dec 12 2007  Simon Wilkinson ***@***.***> 1.4.5
>    * ...
>    * Sat Sep 10 2005  Derek Atkins ***@***.***> 1.4

For the openafs.spec.in from the openafs git repo, this just described
changes to the rpm packaging, not every change in openafs. (I think this
is how it's "supposed" to be used, but I'm not so familiar with rpm
packaging, so I could be wrong.) So this would mention if, for example,
we changed where files were installed, or changed what files were
packaged, etc.

For cellcc, we haven't really changed anything significant about the
packaging, so there's nothing there. So I think that's accurate.

It's also more useful if someone makes local changes to the packaging
(and adds patches or something); that can be documented in some way in
there.

> - Regarding separate spec-files
>    I have no problems with that. I only thought it is more simple and 
> currently not needed with the mentioned platforms.
>    So if you like I can take care to create the  needed spec-files for 
> those platforms.

But it looks like there are no changes needed, right? So we don't need
separate spec files, at least right now.

> - Missing requirements
>    Today I have tried again to build and install cellcc on fresh 
> installed platforms: rhel7, rhel8, fedora34
>    I figure out that some packages are not installed automatically with 
> the current spec-file. I was able to fix that by adding the following 
> lines in the spec-file.
>    (This was working for all those platforms.)
>    - Requires: perl-DBIx-Simple
>    - Requires: perl-DateTime-Format-MySQL
>    - Requires: perl-Log-Log4perl
>    - Requires: perl-Parallel-ForkManager
>    - Requires: perl-Test-Spelling
>    - Requires: perl-Test-Pod
>    - Requires: perl-Filesys-Df
>    I can create a new PR if you like after the current one is merged.

Most of these should be generated automatically from
%{?perl_default_filter}

And I see them in "rpm -qp cellcc-1.21-1.el7.noarch.rpm --requires". Is
this for all platforms? Do you not see any perl module deps at all?

> - Other things that come up
>    There are some preparations needed (not mentioned in the INSTALL 
> file) to be able to build cellcc
>    - yum install "perl(ExtUtils::MakeMaker)" rpm-build git
>      Question: Should this be mentioned in the INSTALL file?

This is a BuildRequires in the spec. There's a tool, yum-builddep, to
install the build deps.

>    - install afs-command
>      Question: Should we add the repo address 
> https://github.com/openafs-contrib/afs-command in the INSTALL file?
>      Or even better, should we add some packaging to afs-command similar 
> like in cellcc? With that this could be addressed with an additional 
> Requires section in the spec-file for cellcc.

Well, it's also in CPAN, where it seems it will never get updated. We
could mention the openafs-contrib project, though, yes.

Packaging for AFS::Command is outside of scope here; that could be
discussed somewhere in the afs-command project.

-- 
Andrew Deason
***@***.***

**Comment by rbrunckhorst** [comment]
Ok, agreed regarding changelog and only one spec-file at the moment.

On 17 Aug 2021, at 16:57, Andrew Deason wrote:

> On Tue, 17 Aug 2021 01:20:48 -0700
> Ralf Brunckhorst ***@***.***> wrote:
>
>> - Regarding changelog:
>>    I thought it would be useful to have the changelog info also in 
>> the
>> package. I was use to it when working with the openSuSE openAFS src
>> package where those changes are included in the spec-file.
>>    That was also true in the openafs git repo for Redhat in the past:
>>    * Wed Dec 12 2007  Simon Wilkinson ***@***.***> 1.4.5
>>    * ...
>>    * Sat Sep 10 2005  Derek Atkins ***@***.***> 1.4
>
> For the openafs.spec.in from the openafs git repo, this just described
> changes to the rpm packaging, not every change in openafs. (I think 
> this
> is how it's "supposed" to be used, but I'm not so familiar with rpm
> packaging, so I could be wrong.) So this would mention if, for 
> example,
> we changed where files were installed, or changed what files were
> packaged, etc.
>
> For cellcc, we haven't really changed anything significant about the
> packaging, so there's nothing there. So I think that's accurate.
>
> It's also more useful if someone makes local changes to the packaging
> (and adds patches or something); that can be documented in some way in
> there.
>
>> - Regarding separate spec-files
>>    I have no problems with that. I only thought it is more simple and
>> currently not needed with the mentioned platforms.
>>    So if you like I can take care to create the  needed spec-files 
>> for
>> those platforms.
>
> But it looks like there are no changes needed, right? So we don't need
> separate spec files, at least right now.
>
>> - Missing requirements
>>    Today I have tried again to build and install cellcc on fresh
>> installed platforms: rhel7, rhel8, fedora34
>>    I figure out that some packages are not installed automatically 
>> with
>> the current spec-file. I was able to fix that by adding the following
>> lines in the spec-file.
>>    (This was working for all those platforms.)
>>    - Requires: perl-DBIx-Simple
>>    - Requires: perl-DateTime-Format-MySQL
>>    - Requires: perl-Log-Log4perl
>>    - Requires: perl-Parallel-ForkManager
>>    - Requires: perl-Test-Spelling
>>    - Requires: perl-Test-Pod
>>    - Requires: perl-Filesys-Df
>>    I can create a new PR if you like after the current one is merged.
>
> Most of these should be generated automatically from
> %{?perl_default_filter}
>
> And I see them in "rpm -qp cellcc-1.21-1.el7.noarch.rpm --requires". 
> Is
> this for all platforms? Do you not see any perl module deps at all?
>
That is working in rhel7:
Output from rpmbuild: Requires: /usr/bin/env perl >= 0:5.008_000 
perl(AFS::CellCC) perl(AFS::CellCC::CLI) 
perl(AFS::CellCC::CLI::ccc_debug) perl(AFS::CellCC::CLI::cellcc) 
perl(AFS::CellCC::Check) perl(AFS::CellCC::Config) 
perl(AFS::CellCC::Const) perl(AFS::CellCC::DB) perl(AFS::CellCC::Dump) 
perl(AFS::CellCC::Log) perl(AFS::CellCC::Remote) 
perl(AFS::CellCC::Remoteclient) perl(AFS::CellCC::Restore) 
perl(AFS::CellCC::Util) perl(AFS::CellCC::VOS) perl(Carp) perl(Cwd) 
perl(DBI) perl(DBIx::Simple) perl(Data::Dumper) perl(DateTime) 
perl(DateTime::Format::MySQL) perl(DateTime::Format::Strptime) 
perl(Digest) perl(Errno) perl(Exporter) perl(File::Basename) 
perl(File::Copy) perl(File::Spec) perl(File::Temp) perl(File::stat) 
perl(Filesys::Df) perl(Getopt::Long) perl(IO::Handle) perl(JSON::PP) 
perl(Log::Log4perl) perl(Net::Domain) perl(POSIX) 
perl(Parallel::ForkManager) perl(strict) perl(warnings)

***@***.*** cellcc]$ rpm -qp 
packaging/RPMS/noarch/cellcc-1.21-1.el7.noarch.rpm --requires
/usr/bin/env
config(cellcc) = 1.21-1.el7
kstart
perl >= 0:5.008_000
perl(:MODULE_COMPAT_5.16.3)
perl(AFS::CellCC)
perl(AFS::CellCC::CLI)
perl(AFS::CellCC::CLI::ccc_debug)
perl(AFS::CellCC::CLI::cellcc)
perl(AFS::CellCC::Check)
perl(AFS::CellCC::Config)
perl(AFS::CellCC::Const)
perl(AFS::CellCC::DB)
perl(AFS::CellCC::Dump)
perl(AFS::CellCC::Log)
perl(AFS::CellCC::Remote)
perl(AFS::CellCC::Remoteclient)
perl(AFS::CellCC::Restore)
perl(AFS::CellCC::Util)
perl(AFS::CellCC::VOS)
perl(Carp)
perl(Cwd)
perl(DBI)
perl(DBIx::Simple)
perl(Data::Dumper)
perl(DateTime)
perl(DateTime::Format::MySQL)
perl(DateTime::Format::Strptime)
perl(Digest)
perl(Errno)
perl(Exporter)
perl(File::Basename)
perl(File::Copy)
perl(File::Spec)
perl(File::Temp)
perl(File::stat)
perl(Filesys::Df)
perl(Getopt::Long)
perl(IO::Handle)
perl(JSON::PP)
perl(Log::Log4perl)
perl(Net::Domain)
perl(POSIX)
perl(Parallel::ForkManager)
perl(strict)
perl(warnings)
perl-DBD-MySQL
remctl
rpmlib(CompressedFileNames) <= 3.0.4-1
rpmlib(FileDigests) <= 4.6.0-1
rpmlib(PayloadFilesHavePrefix) <= 4.0-1
rpmlib(PayloadIsXz) <= 5.2-1

but in rhel8:
Output from rpmbuild: Requires: /usr/bin/perl

rpm -qp packaging/RPMS/noarch/cellcc-1.21-1.el8.noarch.rpm --requires
/usr/bin/perl
config(cellcc) = 1.21-1.el8
kstart
perl(:MODULE_COMPAT_5.26.3)
perl-DBD-MySQL
remctl
rpmlib(CompressedFileNames) <= 3.0.4-1
rpmlib(FileDigests) <= 4.6.0-1
rpmlib(PayloadFilesHavePrefix) <= 4.0-1
rpmlib(PayloadIsXz) <= 5.2-1

And in fedora34:
Output from rpmbuild: Requires: /usr/bin/perl

***@***.*** cellcc]$ rpm -qp 
packaging/RPMS/noarch/cellcc-1.21-1.fc34.noarch.rpm --requires
/usr/bin/perl
config(cellcc) = 1.21-1.fc34
kstart
perl(:MODULE_COMPAT_5.32.1)
perl-DBD-MySQL
remctl
rpmlib(CompressedFileNames) <= 3.0.4-1
rpmlib(FileDigests) <= 4.6.0-1
rpmlib(PayloadFilesHavePrefix) <= 4.0-1
rpmlib(PayloadIsZstd) <= 5.4.18-1

That was the reason I have added those Requires lines and on rhel7 they 
would not brake something.

>> - Other things that come up
>>    There are some preparations needed (not mentioned in the INSTALL
>> file) to be able to build cellcc
>>    - yum install "perl(ExtUtils::MakeMaker)" rpm-build git
>>      Question: Should this be mentioned in the INSTALL file?
>
> This is a BuildRequires in the spec. There's a tool, yum-builddep, to
> install the build deps.
>
That is correct.

>>    - install afs-command
>>      Question: Should we add the repo address
>> https://github.com/openafs-contrib/afs-command in the INSTALL file?
>>      Or even better, should we add some packaging to afs-command 
>> similar
>> like in cellcc? With that this could be addressed with an additional
>> Requires section in the spec-file for cellcc.
>
> Well, it's also in CPAN, where it seems it will never get updated. We
> could mention the openafs-contrib project, though, yes.
>
Ok.
> Packaging for AFS::Command is outside of scope here; that could be
> discussed somewhere in the afs-command project.
>
Ok.

**Comment by rbrunckhorst** [comment]
Hi again,

I have found the reason why the perl-requirements for the package did 
not work on rhel8 and fedora34.
There is a package needed to create those RPM Perl dependencies. It is 
called perl-generators

After installing the package the dependency check is working and the 
produced cellcc rpm looks good:
Provides: cellcc = 1.21-1.el8 config(cellcc) = 1.21-1.el8 
perl(AFS::CellCC) perl(AFS::CellCC::CLI) 
perl(AFS::CellCC::CLI::ccc_debug) perl(AFS::CellCC::CLI::cellcc) 
perl(AFS::CellCC::Check) perl(AFS::CellCC::Config) 
perl(AFS::CellCC::Const) = 1.021 perl(AFS::CellCC::DB) 
perl(AFS::CellCC::Dump) perl(AFS::CellCC::Log) perl(AFS::CellCC::Remote) 
perl(AFS::CellCC::Remoteclient) perl(AFS::CellCC::Restore) 
perl(AFS::CellCC::Util) perl(AFS::CellCC::VOS)
Requires(rpmlib): rpmlib(CompressedFileNames) <= 3.0.4-1 
rpmlib(FileDigests) <= 4.6.0-1 rpmlib(PayloadFilesHavePrefix) <= 4.0-1
Requires: /usr/bin/perl perl(:VERSION) >= 5.8.0 perl(AFS::CellCC) 
perl(AFS::CellCC::CLI) perl(AFS::CellCC::CLI::ccc_debug) 
perl(AFS::CellCC::CLI::cellcc) perl(AFS::CellCC::Check) 
perl(AFS::CellCC::Config) perl(AFS::CellCC::Const) perl(AFS::CellCC::DB) 
perl(AFS::CellCC::Dump) perl(AFS::CellCC::Log) perl(AFS::CellCC::Remote) 
perl(AFS::CellCC::Remoteclient) perl(AFS::CellCC::Restore) 
perl(AFS::CellCC::Util) perl(AFS::CellCC::VOS) perl(Carp) perl(Cwd) 
perl(DBI) perl(DBIx::Simple) perl(Data::Dumper) perl(DateTime) 
perl(DateTime::Format::MySQL) perl(DateTime::Format::Strptime) 
perl(Digest) perl(Errno) perl(Exporter) perl(File::Basename) 
perl(File::Copy) perl(File::Spec) perl(File::Temp) perl(File::stat) 
perl(Filesys::Df) perl(Getopt::Long) perl(IO::Handle) perl(JSON::PP) 
perl(Log::Log4perl) perl(Net::Domain) perl(POSIX) 
perl(Parallel::ForkManager) perl(strict) perl(warnings)

***@***.*** cellcc]$ rpm -qp 
packaging/RPMS/noarch/cellcc-1.21-1.el8.noarch.rpm --requires
/usr/bin/perl
config(cellcc) = 1.21-1.el8
kstart
perl(:MODULE_COMPAT_5.26.3)
perl(:VERSION) >= 5.8.0
perl(AFS::CellCC)
perl(AFS::CellCC::CLI)
perl(AFS::CellCC::CLI::ccc_debug)
perl(AFS::CellCC::CLI::cellcc)
perl(AFS::CellCC::Check)
perl(AFS::CellCC::Config)
perl(AFS::CellCC::Const)
perl(AFS::CellCC::DB)
perl(AFS::CellCC::Dump)
perl(AFS::CellCC::Log)
perl(AFS::CellCC::Remote)
perl(AFS::CellCC::Remoteclient)
perl(AFS::CellCC::Restore)
perl(AFS::CellCC::Util)
perl(AFS::CellCC::VOS)
perl(Carp)
perl(Cwd)
perl(DBI)
perl(DBIx::Simple)
perl(Data::Dumper)
perl(DateTime)
perl(DateTime::Format::MySQL)
perl(DateTime::Format::Strptime)
perl(Digest)
perl(Errno)
perl(Exporter)
perl(File::Basename)
perl(File::Copy)
perl(File::Spec)
perl(File::Temp)
perl(File::stat)
perl(Filesys::Df)
perl(Getopt::Long)
perl(IO::Handle)
perl(JSON::PP)
perl(Log::Log4perl)
perl(Net::Domain)
perl(POSIX)
perl(Parallel::ForkManager)
perl(strict)
perl(warnings)
perl-DBD-MySQL
remctl
rpmlib(CompressedFileNames) <= 3.0.4-1
rpmlib(FileDigests) <= 4.6.0-1
rpmlib(PayloadFilesHavePrefix) <= 4.0-1
rpmlib(PayloadIsXz) <= 5.2-1

I don't know why this package is not installed in the vagrant box 
generic/centos8|fedora34 I have used for my testing.

/Ralf

**Comment by adeason** [comment]
On Tue, 17 Aug 2021 11:20:45 -0700
Ralf Brunckhorst ***@***.***> wrote:

> I have found the reason why the perl-requirements for the package did 
> not work on rhel8 and fedora34.
> There is a package needed to create those RPM Perl dependencies. It is 
> called perl-generators

Yeah, apparently we're supposed to BuildRequires on that:

https://fedoraproject.org/wiki/PackagingDrafts/Packaging:Perl/BuildRequires#Build_Requires

That was always available in RHEL7 and earlier, but was split into a
separate package in RHEL8 (and similar Fedora), to allow for
installations to exist without perl, it looks like.

That macro should really fail instead of expanding to nothing when it's
not available; it looks like just omitting the '?' does that (I'm pretty
sure that's what it does, but I can't find _any_ rpm documentation that
actually says that).

-- 
Andrew Deason
***@***.***

**Comment by rbrunckhorst** [comment]
I have tested the build with the additional lines:
BuildRequires:  perl-generators
%perl_default_filter
and this works fine on rhel8 and fedora34.

In RHEL<=7 those perl-requirement routines where included in rpm-build
The perl-generators package for RHEL7 in epel is just a dummy package 
for compatibility.

Regarding the macro:
I have tested without the BuildRequires:  perl-generators and with:
%perl_default_filter
or
%{perl_default_filter}
But it never failed when perl-generators package is not installed.
It builds the package without the required perl modules info.

**Comment by adeason** [comment]
On Tue, 17 Aug 2021 22:44:11 -0700
Ralf Brunckhorst ***@***.***> wrote:

> Regarding the macro:
> I have tested without the BuildRequires:  perl-generators and with:
> %perl_default_filter
> or
> %{perl_default_filter}
> But it never failed when perl-generators package is not installed.
> It builds the package without the required perl modules info.

Indeed, I misunderstood what this macro was doing. The generated
dependencies don't come from the %perl_default_filter macro; that macro
is always around, and just excludes some stuff from the
dependency-generating logic. The generated perl deps just happen
automatically (they still happen even when that macro isn't there), as
long as perl-generators is installed.

So I've removed this bit of it. I've re-pushed the commits, so see if
those seem okay?

-- 
Andrew Deason
***@***.***

**Comment by rbrunckhorst** [comment]
All looks good now.

I have also make a test again for RHEL6, perl-generators package is also 
available in epel-repo as dummy package.
So with the new updates cellcc has still support for this old platform. 
Not sure if this should be mentioned.

 From my side, please merge the changes into master.

BTW: Since today perl-DBIx-Simple-1.37-12.el8 
perl-DBIx-XHTML_Table-1.49-17.el8 perl-SQL-Interp-1.27-1.el8 are now 
available in the official epel-8 repo.

/Ralf

**Comment by adeason** [comment]
On Tue, 17 Aug 2021 01:20:48 -0700
Ralf Brunckhorst ***@***.***> wrote:

>    - install afs-command
>      Question: Should we add the repo address 
> https://github.com/openafs-contrib/afs-command in the INSTALL file?

I meant to add this earlier, but forgot. I put in a small note about
this now.

-- 
Andrew Deason
***@***.***

**Review by adeason — APPROVED** [review]

**Pull request merged**
